optimize getComponent performance

PPpro opened the pull request:
Re: #2394 

Changelog:
在getComponent 时，component.sealed===true（不存在子类）情况下，将 instanceof class 改为 constructor compare  
性能对比：
在component.sealed === false 时，instanceof class 的效率会高一点（只高了6%）
在component.sealed === true 时，constructor compare  的效率会明显高于 instanceof class （20%~40%）

jareguo commented on cocos2d/core/platform/CCClass.js:
```diff
@@ -922,6 +922,11 @@ function CCClass (options) {
         name = cc.js.getClassName(cls);
     }
 
+    cls.sealed = true;
```
麻烦改成 `_sealed`，因为我们不希望用户使用这个属性

jareguo commented on cocos2d/core/utils/base-node.js:
```diff
@@ -51,20 +51,40 @@ function getConstructor(typeOrClassName) {
-        var comp = node._components[i];
-        if (comp instanceof constructor) {
-            return comp;
+    if (constructor._sealed === true) {
```
不好意思，刚刚漏看了，请把 === true 去掉。直接 if 就好，这样比较快~

jareguo commented on cocos2d/core/utils/base-node.js:
```diff
@@ -51,20 +51,40 @@ function getConstructor(typeOrClassName) {
-        if (comp instanceof constructor) {
-            return comp;
+    if (constructor._sealed) {
+        for (var i = 0; i < node._components.length; ++i) {
```
不好意思，又看漏了。既然下面用的是 let，这边也用 let 吧。一个 function 两个 var i 确实比较不好。

PPpro committed:
optimize getComponent performance


optimize getComponent performance


optimize getComponent performance


optimize getComponent performance
